server/hwapi/data_models/repository.py

```diff
@@ -107,3 +111,253 @@ def get_or_create(
+            f"No matching release found for codename '{os.codename}', version '{os.version}'"
+        )
+    return release
+
+
+def clean_vendor_name(name):
+    """Remove "Inc"/"Inc." substring from vendor name and leading whitespaces"""
+    return name.replace("Inc.", "").replace("Inc", "").strip()
+
+
+def get_board_by_validator_data(
+    db: Session, board_validator: device_validators.BoardValidator
+) -> models.Device | None:
+    """Return device object (category==BOARD) matching given Board data"""
+    board = (
+        db.query(models.Device)
+        .join(models.Vendor)
+        .filter(
+            and_(
+                models.Vendor.name.in_(
+                    [
+                        board_validator.manufacturer,
+                        clean_vendor_name(board_validator.manufacturer),
+                    ]
+                ),
+                models.Device.name == board_validator.product_name,
+                models.Device.version == board_validator.version,
+                models.Device.category.in_(
+                    [DeviceCategory.BOARD.value, DeviceCategory.OTHER.value]
+                ),
+            )
+        )
+        .first()
+    )
+    return board
+
+
+def get_devices_by_machine_ids(db: Session, machine_ids: list[int]) -> Query:
+    """
+    Return query that contains Devices joined with Reports and Certificates for a
+    given list of machine IDs
+    """
+    return (
+        db.query(models.Device)
+        .join(
+            models.device_report_association,
+            models.Device.id == models.device_report_association.c.device_id,
+        )
+        .join(
+            models.Report,
+            models.device_report_association.c.report_id == models.Report.id,
+        )
+        .join(models.Certificate, models.Report.certificate_id == models.Certificate.id)
+        .filter(models.Certificate.machine_id.in_(machine_ids))
+    )
+
+
+def find_matching_processor(
+    db: Session,
+    devices_query: Query,
+    cpu_validator: device_validators.ProcessorValidator,
+) -> models.Device | None:
+    """
+    Find a CPU device across the given list of devices that contains
+    model name.
+    """
+    return devices_query.filter(
+        and_(
+            models.Device.name.like(f"%{cpu_validator.version}%"),
+            models.Device.category == DeviceCategory.PROCESSOR,
+        ),
+    ).first()
+
+
+def find_matching_gpu(
+    db: Session, devices_query: Query, gpu_validator: device_validators.GPUValidator
+) -> models.Device | None:
+    """
+    Find a GPU device across the given list of devices that have the same model name and
+    identifier
+    """
+    return devices_query.filter(
+        and_(
+            models.Device.name == gpu_validator.version,
+            models.Device.identifier == gpu_validator.identifier.lower(),
+            models.Device.category.in_(
+                [DeviceCategory.VIDEO.value, DeviceCategory.OTHER]
+            ),
```

Comment:
I see some places where you check that device category is either what's expected or OTHER and other places where you don't check for OTHER. Is there any reason for that? if there is then we need to be more explicit with the reason.

Also it seems to me that OTHER category should be named UNKNOWN instead as it's not really OTHER if it's used this way

Reply:
this logic has been removed